pkg/build/golang/docker.go

```diff
@@ -164,28 +182,88 @@ func (b *DockerGoBuilder) Build(opts *api.BuildInput) (*api.BuildOutput, error)
 		return nil, err
 	}
 
-	args := map[string]*string{
-		"GO_VERSION":        &cfg.GoVersion,
-		"GO_IPFS_VERSION":   &cfg.GoIPFSVersion,
-		"TESTPLAN_EXEC_PKG": &cfg.ExecPkg,
+	opts := types.ImageBuildOptions{
+		Tags: []string{id, in.BuildID},
+		BuildArgs: map[string]*string{
+			"GO_VERSION":        &cfg.GoVersion,
+			"GO_IPFS_VERSION":   &cfg.GoIPFSVersion,
+			"TESTPLAN_EXEC_PKG": &cfg.ExecPkg,
+		},
 	}
 
-	buildOpts := types.ImageBuildOptions{
-		Tags:      []string{id},
-		BuildArgs: args,
-	}
-
-	resp, err := cli.ImageBuild(ctx, tar, buildOpts)
+	// Build the image.
+	resp, err := cli.ImageBuild(ctx, tar, opts)
 	if err != nil {
 		return nil, err
 	}
 	defer resp.Body.Close()
 
+	// Pipe the docker output to stdout.
 	if err := util.PipeDockerOutput(resp.Body, os.Stdout); err != nil {
 		return nil, err
 	}
 
-	return &api.BuildOutput{ArtifactPath: id}, nil
+	out := &api.BuildOutput{ArtifactPath: in.BuildID}
+
+	if cfg.PushRegistry {
+		err := b.pushToRegistry(ctx, cli, in, out)
+		return out, err
+	}
+
+	return out, nil
+}
+
+func (b *DockerGoBuilder) pushToRegistry(ctx context.Context, client *client.Client, in *api.BuildInput, out *api.BuildOutput) error {
+	cfg := in.BuildConfig.(*DockerGoBuilderConfig)
+
+	switch cfg.RegistryType {
+	case "aws":
+		goto AWS
+	default:
+		return fmt.Errorf("no registry type specified, or unrecognised value: %s", cfg.RegistryType)
+	}
+
+AWS:
```

Comment:
1. No need for this goto.
2. 

```go
if cfg.RegistryType != "aws" {
  return ...
}

Reply:
I was putting in some boilerplate for future registry types, and experimenting with idioms for switching without nesting large code blocks. But OK :P